src/manifolds/MultinomialDoublyStochastic.jl

```diff
@@ -0,0 +1,286 @@
+    > doi: [10.1109/tsp.2019.2946024](http://doi.org/10.1109/tsp.2019.2946024),
+    > arXiv: [1802.02628](https://arxiv.org/abs/1802.02628).
+"""
+struct MultinomialDoubleStochasticMatrices{N} <:
+       AbstractEmbeddedManifold{ℝ,DefaultEmbeddingType} where {N} end
+
+function MultinomialDoubleStochasticMatrices(n::Int)
+    return MultinomialDoubleStochasticMatrices{n}()
+end
+
+@doc raw"""
+    check_manifold_point(M::MultinomialDoubleStochasticMatrices, p)
+
+Checks whether `p` is a valid point on the [`MultinomialDoubleStochasticMatrices`](@ref)`(m,n)` `M`,
+i.e. is a  matrix with positive entries whose rows and columns sum to one.
+"""
+check_manifold_point(::MultinomialDoubleStochasticMatrices, ::Any)
+function check_manifold_point(
+    M::MultinomialDoubleStochasticMatrices{n},
+    p;
+    kwargs...,
+) where {n}
+    mpv =
+        invoke(check_manifold_point, Tuple{supertype(typeof(M)),typeof(p)}, M, p; kwargs...)
+    mpv === nothing || return mpv
+    c = sum(p, dims = 1) # due to symmetry we only have to check the cols
+    if !isapprox(norm(c - ones(1, n)), 0.0; kwargs...)
+        return DomainError(
+            c,
+            "The point $(p) does not lie on $M, since its columns do not sum up to one.",
+        )
+    end
+    r = sum(p, dims = 2) # due to symmetry we only have to check the cols
+    if !isapprox(norm(r - ones(n, 1)), 0.0; kwargs...)
+        return DomainError(
+            r,
+            "The point $(p) does not lie on $M, since its rows do not sum up to one.",
+        )
+    end
+    if !(minimum(p) > 0)
+        return DomainError(
+            minimum(p),
+            "The point $(p) does not lie on $M, since at least one of its entries is nonpositive.",
+        )
+    end
+    return nothing
+end
+@doc raw"""
+    check_tangent_vector(M::MultinomialDoubleStochasticMatrices p, X; check_base_point = true, kwargs...)
+
+Checks whether `X` is a valid tangent vector to `p` on the [`MultinomialDoubleStochasticMatrices`](@ref) `M`.
+This means, that `p` is valid, that `X` is of correct dimension and sums to zero along any
+column or row.
+
+The optional parameter `check_base_point` indicates, whether to call
+[`check_manifold_point`](@ref check_manifold_point(::MultinomialDoubleStochasticMatrices, ::Any))  for `p`.
+"""
+function check_tangent_vector(
+    M::MultinomialDoubleStochasticMatrices{n},
+    p,
+    X;
+    check_base_point = true,
+    kwargs...,
+) where {n}
+    if check_base_point
+        mpe = check_manifold_point(M, p; kwargs...)
+        mpe === nothing || return mpe
+    end
+    mpv = invoke(
+        check_tangent_vector,
+        Tuple{supertype(typeof(M)),typeof(p),typeof(X)},
+        M,
+        p,
+        X;
+        check_base_point = false, # already checked above
+        kwargs...,
+    )
+    mpv === nothing || return mpv
+    c = sum(X, dims = 1) # due to symmetry, we only have to check columns
+    if !isapprox(norm(c), 0.0; kwargs...)
```

Comment:
Shouldn't you also check that `X` is symmetric?

Reply:
Now I do that again – but omit something else ;)